### PR TITLE
Feature difference

### LINKED ISSUE #2: Difference
## Behavior

> What the function does - inputs and outputs

- **The Function**
  - [ ] The function's name is clear and helpful
  - [ ] The parameter names are clear and helpful
  - [ ] Default parameters are used if possible
- **JSDoc Comment**
  - [ ] The docstring is well-formatted
  - [ ] There are no spelling mistakes
  - [ ] The description of the function's behavior is clear
  - [ ] All parameters are documented with types
  - [ ] Documented parameter names match the function's parameter names
  - [ ] Parameters have a good description
  - [ ] If default parameters are used, they are the correct type
  - [ ] The return value is documented with a type
  - [ ] The return value has a clear description
- **README**
  - [ ] The documentation is rendered into the README
  - [ ] Docs were generated just before pushing
  - [ ] Docs in the README matches the JSDoc comment
  - [ ] The README contains at least one use case
- **Unit Tests**
  - [ ] Unit tests are present
  - [ ] There are at least 8 test cases
  - [ ] All `describe`s and `it`s have clear and helpful names
  - [ ] All test names actually describe what is being tested
  - [ ] Tests are grouped logically using `describe`
  - [ ] There is only one `expect` inside every `it`
  - [ ] There is no logic (loops, conditionals, ...) in the test cases
  - [ ] Side-effects are tested, if necessary
  - [ ] Tests were run just before pushing
  - [ ] All tests pass (`file-name.pass.txt`)
  - [ ] Tricky edge cases are tested (extra credit)

## Strategy

> The logic behind the solution

- **The Function**
  - [ ] Variables have names that help understand the strategy
  - [ ] Spacing and formatting help to understand the strategy
  - [ ] The code is written as simply and clearly as possible
  - [ ] Comments _are_ used to label important goals in your strategy
  - [ ] Comments _are not_ used to describe how the code works
- **README**
  - [ ] There is an explanation of the solution's strategy in the README
  - [ ] The explanation is clear and understandable
  - [ ] The explanation actually matches the function that was written

## Implementation

> The code used to write the solution

- **Formatting**
  - [ ] The code was formatted just before pushing
  - [ ] No errors occurred while formatting the code
- **Linting**
  - [ ] There should be no linting errors or warnings in the `file-name.lint.txt` file
  - [ ] Spelling should be correct in all code, comments and docstrings
  - [ ] The code was linted just before pushing
- **The solution file**
  - [ ] Has a JSDoc comment
  - [ ] Has an exported function
  - [ ] Does no other code or comments (you can use the sandbox for saving extra ideas)
- **The function body _Should_ ...**
  - [ ] Be well and consistently formatted
  - [ ] Have clear and helpful variable names
  - [ ] Declare variables in the smallest scope possible
  - [ ] Be a simple and readable implementation
  - [ ] Use short comments to describe the strategy
  - [ ] Use short comments to describe important language feature choices
- **The function body should _Not_ ...**
  - [ ] Have extra comments that are not about strategy or important decisions
  - [ ] Have any user interactions (`prompt`, `alert`, `confirm`)
  - [ ] Have any `console` calls (`.log`, `.assert`, ...)
  - [ ] Have any `debugger` statements

### PR DESCRIPTION
**Pull request by erkamguresen**
---
name: solution code review challenge: difference 
about: difference PR for code review, complete with checklist
title: 'Code Review: difference, additive'
---

Before the reviewer merges this PR to main/master, all of these boxes must be checked. Only the reviewer can check boxes.

## Behavior

> What the function does - inputs and outputs

- **The Function**
  - [x] The function's name is clear and helpful
  - [x] The parameter names are clear and helpful
  - [x] Default parameters are used if possible
- **JSDoc Comment**
  - [x] The docstring is well-formatted
  - [x] There are no spelling mistakes
  - [ ] The description of the function's behavior is clear
  - [x] All parameters are documented with types
  - [x] Documented parameter names match the function's parameter names
  - [x] Parameters have a good description
  - [x] If default parameters are used, they are the correct type
  - [x] The return value is documented with a type
  - [x] The return value has a clear description
- **README**
  - [x] The documentation is rendered into the README
  - [x] Docs were generated just before pushing
  - [x] Docs in the README matches the JSDoc comment
  - [x] The README contains at least one use case
- **Unit Tests**
  - [x] Unit tests are present
  - [x] There are at least 8 test cases
  - [x] All `describe`s and `it`s have clear and helpful names
  - [x] All test names actually describe what is being tested
  - [x] Tests are grouped logically using `describe`
  - [x] There is only one `expect` inside every `it`
  - [x] There is no logic (loops, conditionals, ...) in the test cases
  - [x] Side-effects are tested, if necessary
  - [x] Tests were run just before pushing
  - [x] All tests pass (`file-name.pass.txt`)
  - [ ] Tricky edge cases are tested (extra credit)

## Strategy

> The logic behind the solution

- **The Function**
  - [x] Variables have names that help understand the strategy
  - [x] Spacing and formatting help to understand the strategy
  - [x] The code is written as simply and clearly as possible
  - [ ] Comments _are_ used to label important goals in your strategy
  - [ ] Comments _are not_ used to describe how the code works
- **README**
  - [x] There is an explanation of the solution's strategy in the README
  - [ ] The explanation is clear and understandable
  - [ ] The explanation actually matches the function that was written

## Implementation

> The code used to write the solution

- **Formatting**
  - [x] The code was formatted just before pushing
  - [x] No errors occurred while formatting the code
- **Linting**
  - [x] There should be no linting errors or warnings in the `file-name.lint.txt` file
  - [x] Spelling should be correct in all code, comments and docstrings
  - [x] The code was linted just before pushing
- **The solution file**
  - [x] Has a JSDoc comment
  - [x] Has an exported function
  - [x] Does no other code or comments (you can use the sandbox for saving extra ideas)
- **The function body _Should_ ...**
  - [x] Be well and consistently formatted
  - [x] Have clear and helpful variable names
  - [x] Declare variables in the smallest scope possible
  - [x] Be a simple and readable implementation
  - [x] Use short comments to describe the strategy
  - [x] Use short comments to describe important language feature choices
- **The function body should _Not_ ...**
  - [x] Have extra comments that are not about strategy or important decisions
  - [ ] Have any user interactions (`prompt`, `alert`, `confirm`)
  - [ ] Have any `console` calls (`.log`, `.assert`, ...)
  - [ ] Have any `debugger` statements

### CODE REVIEW

File: src/difference/difference_solution1/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+```js
+difference([1, 2, 1], [2, 3]); //  [1]
+```
+
+> Docs generated: Wed May 19 2021, 1:46:52 AM
+
+
+## Strategy
+
+A additive approach is chosen to find the difference between two arrays. Starting from an empty array and added all the different elements to the array.
```
Comment (rago89) [inline src/difference/difference_solution1/README.md]: I see the strategy very empty, I see few arguments to establish an implementation according to the Strategy

File: src/difference/difference_solution1/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+];
+
+const shoppingCart = [
+  { name: "teddy bear", price: 3 },
+  { name: "milk", price: 1 }
+];
+const stillWishedItems = arrayDifference(whishList, shoppingCart);
+```
+
+
```
Comment (rago89) [inline src/difference/difference_solution1/README.md]: Is good to delete all the empty spaces in the document to have a  more clean document

File: src/difference/difference_solution1/difference.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @example
+ *
+ * difference([1, 2, 1], [2, 3]); //  [1]
+ */
+export const arrayDifference = (
+  arrayToFindTheDifference = [],
+  arrayToCompare = [],
+) => {
+  // ensure that a new array is created
```
Comment (rago89) [inline src/difference/difference_solution1/difference.js:27]: To many comments, as a coder you only need comments only in the places that you may think is going to be tricky to debug for future developers, as coaches say to many comments give the impression that the code is hard to understand

File: src/difference/difference_solution1/difference.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const returnArray = [];
+  // ensure that every element in the array is checked
+  for (let i = 0; i < arrayToFindTheDifference.length; i++) {
```
Comment (rago89) [inline src/difference/difference_solution1/difference.js:30]: When you are going to iterate only on elements, and you don't need the indexes a `for of loop` is more compact and more easy to read

File: src/difference/difference_solution1/difference.spec.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    it('should not contain [3]', () => {
+      const actual = arrayDifference([2, 1], [2, 3, 4]);
+      expect(actual).not.toContain(3);
+    });
+    it('should not contain [4]', () => {
+      const actual = arrayDifference([2, 1], [2, 3, 4, 4]);
+      expect(actual).not.toContain(4);
+    });
+  });
+});
```
Comment (rago89) [inline src/difference/difference_solution1/difference.spec.js:96]: Are missing tricky edge cases are tested, see more tests with different types of values to see the behavior, or something different from only  just integers

File: src/difference/difference_solution1/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+A additive approach is chosen to find the difference between two arrays. Starting from an empty array and added all the different elements to the array.
+
+---
+
+## Implementation
+
+Solution starts with an empty array . 
+
+Since the elements of first array which are not in the second array are  wanted, all the elements in the first array is searched one by one.
```
Comment (rago89) [inline src/difference/difference_solution1/README.md]: In the implementation is missing language features and which lines of code can you use to make your strategy a reality?

File: src/difference/difference_solution1/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
+<!-- BEGIN DOCS -->
+# Difference: finds the numbers that are in the first array but not in the second array
```
Comment (rago89) [inline src/difference/difference_solution1/README.md:3]: Which numbers is going to find? I see more like a function that compares the values of one array with another and returns the value that differs from the first array with the second array

**Review by rago89 — CHANGES_REQUESTED** [review]
Hello, Here are some reviews and suggestions.

File: src/difference/difference_solution1/README.md (hunk 7 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/README.md:3]: Read me file description for the function is as follows: "Creates an array of values that are in the first array, but not not in the second array."

and the given examples are : "@example
 *
 * difference([2, 1], [2, 3]); //  [1]
 *
 * @example
 *
 * difference([1, 2, 1], [2, 3]); //  [1]"  so as the main rule of the TDD write the least code that passes tests. Thus only number are included and the definition is clearified.

File: src/difference/difference_solution1/README.md (hunk 1 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/README.md]: Which part is do you think empty ?

File: src/difference/difference_solution1/README.md (hunk 6 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/README.md]: " implementation is missing language features and which lines of code can you use to make your strategy" If I do this as I understand it it will look like coding tutorial instead of a documentation

Comment (erkamguresen) [reply src/difference/difference_solution1/README.md]: Can you elaborate?

File: src/difference/difference_solution1/difference.js (hunk 4 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/difference.js:30]: I find this more readable and it more common across programming languages

File: src/difference/difference_solution1/README.md (hunk 1 above)
Comment (rago89) [reply src/difference/difference_solution1/README.md]: Your implementation should be is according to your strategy, In your strategy you specify in a very general way all the steps that you have to follow to implement your code, in the repo of [bsi](https://github.com/HackYourFutureBelgium/behavior-strategy-implementation), you can see templates about strategy that can be a guide to develop yours

File: src/difference/difference_solution1/difference.spec.js (hunk 5 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/difference.spec.js:96]: https://en.wikipedia.org/wiki/Edge_case

In the java doc it is noted that the function is for numbers. Anyone that will use the function should pass numbers. So other input types are not handled in the method unless it is an outside call or operation like ; read write files, databases, user input etc... You cannot have control over these and they can give unacceptable inputs. Shortly if you go that road every function you write, even the smallest ones will start with input checks.

File: src/difference/difference_solution1/difference.js (hunk 4 above)
Comment (rago89) [reply src/difference/difference_solution1/difference.js:30]: This is only as a suggestion, something that is suggested in code is to build a compact and readable code, less code less complications, the for of loop is one example, another is the global object that in each actualization of JS more methods are included to improve the code.

File: src/difference/difference_solution1/difference.js (hunk 3 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/difference.js:27]: JS doc is the documentation and it will be used by intellicence to guide you while you write the code. So the more the better (up to a point). But I totally agree with the comments in the function body, but the check list 
" Comments are used to label important goals in your strategy
 Comments are not used to describe how the code works" 
 forces these kind of things in small methods. Any suggestions how these can be done in this method?

File: src/difference/difference_solution1/README.md (hunk 2 above)
Comment (erkamguresen) [reply src/difference/difference_solution1/README.md]: done

File: src/difference/difference_solution1/README.md (hunk 6 above)
Comment (rago89) [reply src/difference/difference_solution1/README.md]: You can see it as a code tutorial if you want to, the idea if the implementation, is to build how is going to be the process to reach the goal according to the strategy,  in your case which statement are you going to use to search on each element, (a regular for loop, a for of loop, while loop, the  forEach() method, or are you going to use filter or map methods?), to test the condition there are various control structures which could be (if and else, switch, ternary operator). 
Once you finish your implementation you start working in your code according to it. Of course probably is going to happen that you have to modify your implementation because some things in your first approach didn't work .
